pkg/agent/cniserver/ipam/antrea_ipam_controller.go

```diff
@@ -81,21 +99,86 @@ func (c *AntreaIPAMController) Run(stopCh <-chan struct{}) {
 	}()
 
 	klog.InfoS("Starting", "controller", controllerName)
-	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced, c.ipPoolInformer.Informer().HasSynced, c.podInformer.Informer().HasSynced) {
 		return
 	}
 
 	<-stopCh
 }
 
-func (c *AntreaIPAMController) getIPPoolsByNamespace(namespace string) []string {
+func (c *AntreaIPAMController) getIPPoolsByPod(namespace, name string) ([]string, []net.IP, *crdv1a2.IPAddressOwner, error) {
+	// Find IPPool by Pod
+	var IPs []net.IP
+	var preallocatedOwner *crdv1a2.IPAddressOwner
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		klog.Warningf("Get pod %s/%s failed, err=%+v", namespace, name, err)
+		return nil, nil, nil, err
+	}
+	// Collect specified IPs if exist
+	IPStrings, _ := pod.Annotations[AntreaIPAMPodIPAnnotationKey]
+	IPStrings = strings.ReplaceAll(IPStrings, " ", "")
+	var ipErr error
+	if IPStrings != "" {
+		splittedIPStrings := strings.Split(IPStrings, AntreaIPAMAnnotationDelimiter)
+		for _, IPString := range splittedIPStrings {
+			IP := net.ParseIP(IPString)
+			if IPString != "" && IP == nil {
+				ipErr = fmt.Errorf("invalid IP annotation %s", IPStrings)
+				IPs = nil
+				break
+			}
+			IPs = append(IPs, IP)
+		}
+	}
+
+	if persistIPString := pod.Annotations[AntreaIPAMPersistIPAnnotationKey]; persistIPString == AntreaIPAMPersistIPAnnotationValue {
+	ownerReferenceLoop:
+		for _, ownerReference := range pod.OwnerReferences {
+			if ownerReference.Controller != nil && *ownerReference.Controller == true {
+				switch ownerReference.Kind {
+				case "StatefulSet":
+					statefulSetName, index, err := k8s.ParseStatefulSetName(name)
+					if err != nil {
+						klog.Warningf("Invalid StatefulSet name: %s", name)
+						break ownerReferenceLoop
+					}
+					preallocatedOwner = &crdv1a2.IPAddressOwner{StatefulSet: &crdv1a2.StatefulSetOwner{
+						Name:      statefulSetName,
+						Namespace: namespace,
+						Index:     index,
+					}}
+					break ownerReferenceLoop
+				}
+			}
+		}
+	}
+
+	annotations, exists := pod.Annotations[AntreaIPAMAnnotationKey]
```

Comment:
are we expecting IP Pool annotation on the Pod? not StatefulSet?

Reply:
Currently we use annotation on PodTemplate of StatefulSet/Deployment, thus we need to read annotation from Pod, not StatefulSet/Deployment.
This method can reduce 1 additional query for StatefulSet/Deployment, but I'm OK to move the annotation to StatefulSet/Deployment.
@tnqn @jianjuns What's your opinion?

Reply:
Yes, I thought about this. I feel annotation on PodTemplate is ok, but we should document annotations must be the same for replicas. Maybe also add a comments somewhere (I think with the annotation key constants).

Reply:
OK. Let's keep existing implementation.